integration-tests/src/test/java/org/apache/fineract/integrationtests/RefundForActiveLoansWithAdvancedPaymentAllocationTest.java

```diff
@@ -288,15 +288,215 @@ public void refundForActiveLoanWithDefaultPaymentAllocation() {
+                    String.valueOf(penalty), secondInstallmentChargeAddedDate, String.valueOf(penaltyPortion)));
+
+            GetLoansLoanIdResponse loanDetails = loanTransactionHelper.getLoanDetails((long) loanId);
+
+            GetLoansLoanIdRepaymentPeriod firstRepaymentInstallment = loanDetails.getRepaymentSchedule().getPeriods().get(2);
+            GetLoansLoanIdRepaymentPeriod secondRepaymentInstallment = loanDetails.getRepaymentSchedule().getPeriods().get(3);
+            GetLoansLoanIdRepaymentPeriod thirdRepaymentInstallment = loanDetails.getRepaymentSchedule().getPeriods().get(4);
+
+            assertEquals(5, loanDetails.getRepaymentSchedule().getPeriods().size());
+            assertEquals(feePortion, firstRepaymentInstallment.getFeeChargesDue());
+            assertEquals(feePortion, firstRepaymentInstallment.getFeeChargesOutstanding());
+            assertEquals(penaltyPortion, firstRepaymentInstallment.getPenaltyChargesDue());
+            assertEquals(penaltyPortion, firstRepaymentInstallment.getPenaltyChargesOutstanding());
+            assertEquals(250.0f, firstRepaymentInstallment.getPrincipalDue());
+            assertEquals(250.0f, firstRepaymentInstallment.getPrincipalOutstanding());
+            assertEquals(400.0f, firstRepaymentInstallment.getTotalDueForPeriod());
+            assertEquals(400.0f, firstRepaymentInstallment.getTotalOutstandingForPeriod());
+            assertEquals(LocalDate.of(2023, 1, 31), firstRepaymentInstallment.getDueDate());
+
+            assertEquals(feePortion, secondRepaymentInstallment.getFeeChargesDue());
+            assertEquals(feePortion, secondRepaymentInstallment.getFeeChargesOutstanding());
+            assertEquals(penaltyPortion, secondRepaymentInstallment.getPenaltyChargesDue());
+            assertEquals(penaltyPortion, secondRepaymentInstallment.getPenaltyChargesOutstanding());
+            assertEquals(400.0f, secondRepaymentInstallment.getTotalDueForPeriod());
+            assertEquals(400.0f, secondRepaymentInstallment.getTotalOutstandingForPeriod());
+            assertEquals(LocalDate.of(2023, 3, 2), secondRepaymentInstallment.getDueDate());
+
+            assertEquals(0.0f, thirdRepaymentInstallment.getFeeChargesDue());
+            assertEquals(0.0f, thirdRepaymentInstallment.getFeeChargesOutstanding());
+            assertEquals(0.0f, thirdRepaymentInstallment.getPenaltyChargesDue());
+            assertEquals(0.0f, thirdRepaymentInstallment.getPenaltyChargesOutstanding());
+            assertEquals(250.0f, thirdRepaymentInstallment.getTotalDueForPeriod());
+            assertEquals(250.0f, thirdRepaymentInstallment.getTotalOutstandingForPeriod());
+            assertEquals(LocalDate.of(2023, 4, 1), thirdRepaymentInstallment.getDueDate());
+
+            businessDateHelper.updateBusinessDate(new BusinessDateRequest().type(BusinessDateType.BUSINESS_DATE.getName())
+                    .date("2023.03.01").dateFormat("yyyy.MM.dd").locale("en"));
```

Comment:
On this day, you only have one installment which is past due., as the due date of the second installments is 2nd of march.

Reply:
I changed it to 28th of January, here we have multiple installments paid in advance, and still the second installmnet will be fully refunded

Reply:
I suggest to move the time to after 1st of April, so all the installments will become past due. And at that point let's make payment. 

0) Create the loan with the same configuration, add the same fees and penalties
1) Set the time to 1st of April -> this makes all the installments past due
2) make pay 100 as repayment, first installment fee outstanding should be 0, penalty should be 50. -> this proves that we are in horizontal payment strategy as on vertical it would have payed the fee's on the 1st and on the 2nd installment. 
3) make another payment 710 -> first and second installment should be fully payed., 3rd installment principal outstanding is 240
4) make refund with 600 -> 
    3rd installment prinipal outstanding will be 250 again. (10)
    2nd installment principal outstanding will be 250         (260)
    2nd installment fee outstanding will be 50           (310)
    2nd instalment's charge outstanding will be 100  (410)
    1st instalment principal outstanding will be 60 (600)
    
    
My understanding that in case of **vertical** the step 4 (refund 600) will look like this: 
   3rd installment principal outstanding will be 250 (10)
   2nd installment principal outstanding will be 250 (260)
   1st installment principal outstanding will be 250 (510)
   2nd installment fee outstanding will be 50 (560)
   1st installment fee outstanding will be 40 (600)
   
  cc: @adamsaghy

Reply:
in the current state of Fineract we don't refund past due installments, never did. The advance payment allocation change covers it but only for future proofment reasons. We can only have 1 due installment, so to test if the horizontal processing works we need to have multiple in advance installments, and that's what the test does.